third_party/proton/csrc/lib/Profiler/RoctracerProfiler.cpp

```diff
@@ -0,0 +1,327 @@
+    externalIdMap[RoctracerProfiler::CorrelationDomain::size];
+
+} // namespace
+
+// External correlation
+void RoctracerProfiler::pushCorrelationID(uint64_t id, CorrelationDomain type) {
+  if (!instance().externalCorrelationEnabled) {
+    return;
+  }
+  externalIdMap[type].push_back(id);
+}
+
+void RoctracerProfiler::popCorrelationID(CorrelationDomain type) {
+  if (!instance().externalCorrelationEnabled) {
+    return;
+  }
+  externalIdMap[type].pop_back();
+}
+
+void RoctracerProfiler::startOp(const Scope &scope) {
+  pushCorrelationID(scope.scopeId, Default);
+}
+
+void RoctracerProfiler::stopOp(const Scope &scope) {
+  popCorrelationID(Default);
+}
+
+void RoctracerProfiler::setOpInProgress(bool value) {
+  roctracerState.isRecording = value;
+}
+
+bool RoctracerProfiler::isOpInProgress() { return roctracerState.isRecording; }
+
+void RoctracerProfiler::doStart() {
+  // Inline Callbacks
+  // roctracer::enableDomainCallback<true>(ACTIVITY_DOMAIN_HSA_API,
+  //                                       api_callback, nullptr);
+  roctracer::enableDomainCallback<true>(ACTIVITY_DOMAIN_HIP_API, apiCallback,
+                                        nullptr);
+
+  // Activity Records
+  roctracer_properties_t properties;
+  memset(&properties, 0, sizeof(roctracer_properties_t));
+  properties.buffer_size = 0x1000;
+  properties.buffer_callback_fun = activityCallback;
+  roctracer::openPool<true>(&properties);
+  roctracer::enableDomainActivity<true>(ACTIVITY_DOMAIN_HIP_OPS);
+  roctracer::start();
+}
+
+void RoctracerProfiler::doFlush() {
+  // Implement reliable flushing.  Wait for all dispatched ops to be reported
+  auto ret = hip::deviceSynchronize<true>();
+  roctracer::flushActivity<true>();
+  std::unique_lock<std::mutex> lock(flushState.mutex_);
+  // Load ending id from the running max
+  auto correlationId = flushState.maxCorrelationId_.load();
+
+  // Poll on the worker finding the final correlation id
+  int timeout = 500;
+  while ((flushState.maxCompletedCorrelationId_ < correlationId) && --timeout) {
+    lock.unlock();
+    roctracer::flushActivity<true>();
+    usleep(1000);
+    lock.lock();
+  }
+}
+
+void RoctracerProfiler::doStop() {
+  roctracer::stop();
+  // roctracer::disable_domain_callback<true>(ACTIVITY_DOMAIN_HSA_API);
+  roctracer::disableDomainCallback<true>(ACTIVITY_DOMAIN_HIP_API);
+  roctracer::disableDomainActivity<true>(ACTIVITY_DOMAIN_HIP_OPS);
+  roctracer::closePool<true>();
+}
+
+void RoctracerProfiler::activityCallback(const char *begin, const char *end,
+                                         void *arg) {
+  RoctracerProfiler &profiler =
+      dynamic_cast<RoctracerProfiler &>(RoctracerProfiler::instance());
+  auto &correlation = profiler.correlation;
+  auto &dataSet = profiler.dataSet;
+
+  std::unique_lock<std::mutex> lock(flushState.mutex_);
+  const roctracer_record_t *record = (const roctracer_record_t *)(begin);
+  const roctracer_record_t *end_record = (const roctracer_record_t *)(end);
```

Comment:
Can you also change all variable names to the camel convention (except those from the roctracer library)? Thanks